File: LC/[0368] Largest Divisible Subset.py
########
# 20240505
########
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        N=len(nums)
        nums.sort()
        dp = [1] * N
        prev = [-1] * N

        for ii in range(N):
            for jj in reversed(range(ii)):
                if nums[ii]%nums[jj]==0:
                    dp[ii]=max(dp[ii], dp[jj]+1)
                    if dp[ii]==dp[jj]+1: # we have a larger subset
                        prev[ii]=jj #record the path for trace back
        

        #figure out which ii has the max length
        mxl=0
        idx=-1
        for ii in range(N):
            if dp[ii] > mxl:
                mxl=dp[ii]
                idx=ii

        res=[]
        while idx != -1:
            res.append(nums[idx])
            idx=prev[idx]

        return res
        
#####################
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        N = len(nums)  # take measurement

        nums.sort()  # sorted in increasing order inplace

        dp = [1]*N  # dp[ii] the size of max divisible subset ending @ ii
        prev = [-1]*N  # record the prev divisor

        lds = 0  # global largest divisible subset
        ldsi = 0  # the index of teh largest divisible subset end
        res = []  # elements of the largest divisible subset

        for ii, nn in enumerate(nums):
            for jj in range(ii):
                if nums[ii] % nums[jj] != 0:
                    continue  # ii is bigger

                if 1+dp[jj] > dp[ii]:
                    dp[ii] = 1+dp[jj]
                    prev[ii] = jj

            if dp[ii] > lds:
                lds = dp[ii]
                ldsi = ii

        ii = ldsi

        while ii != -1:
            res.append(nums[ii])
            ii = prev[ii]
        return res

File: LC/[0368] Largest Divisible Subset.cpp
/********************
 * 20240512
***********************/
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        /*
         + nums[ii]: (distinct) postive int
         -> largest subset
         s.t.: all pairs divisible
         * when we are finding subset->sort

        # idea: 
        * all pairs divisible a/b where a>b
        * if we sort answer
        * answer[ii] divisible answer[ii-1] then answer[ii] can be divisible by all answer[jj]  

        # we are looking for jj < ii and nums[ii]%nums[jj]==0 -> type II DP
        */
        nums.insert(nums.begin(), 0); //insert a dummy header
        int N = nums.size();
        sort(nums.begin(), nums.end());
        vector<int> dp(N,  1);
        vector<int> prev(N,  -1);
        vector<int> answer;
        
        //initialize dp
        dp[0]=0;

        //we need to output path. -> record prev
        /*
            dp[ii] is the largest divisible subset ending at ii
            dp[ii] = max(dp[ii], dp[jj] + 1) where jj < ii and nums[ii]%nums[jj]==0
            prev[ii]=jj
        */

        //find out dp and identify the size of largest divisible subset
        for(int ii=1; ii<N; ++ii){
            for(int jj=ii-1; jj>=1; --jj){
                if(nums[ii]%nums[jj]!=0) continue;//skip those not divisible
                if(dp[jj]+1 > dp[ii]){
                    prev[ii]=jj;
                    dp[ii]=dp[jj]+1;
                }
            }
        }
        //traverse the dp[ii], once found the max_size, trace back for the path
        int mx_sz = *(std::max_element(dp.begin(), dp.end()));

        // cout << " dp: " << endl;
        // for(int ii=0; ii<N; ++ii){
        //     cout << dp[ii] << "->";
        // }
        // cout << endl;
        // cout << " prev: " << endl;
        // for(int ii=0; ii<N; ++ii){
        //     cout << prev[ii] << "->";
        // }
        // cout << endl;
        // cout << " mx_sz: " << mx_sz << endl;
        for(int ii=1; ii<dp.size(); ++ii){ //locate one max size idx
            if(dp[ii]==mx_sz){
                int jj = ii;
                while(jj>=0){
                    answer.push_back(nums[jj]);
                    jj=prev[jj];
                }
                break;
            }
        }

        return answer;
    }
};


/*---------------------------------------------------- */
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int N=nums.size();
        sort(nums.begin(), nums.end()); //remember to sort 
        vector<int> dp(N, 1);     //dp[ii]: considering nums[0:ii], the largest divisible subset ending @ nums[ii]
        vector<int> prev(N, -1);  //whenever DP need to output path -> record the path by prev only, do not record the whole path
        for(int ii=0; ii<N; ++ii){//Type II DP: N^2
            for(int jj=0; jj<ii; ++jj){
                if(nums[ii]%nums[jj]==0){
                    dp[ii] = max(dp[ii], dp[jj]+1);
                    if(dp[ii] == dp[jj]+1){
                        prev[ii]=jj; //record prev for trace back
                    }
                }

            }
        }

        int mxl=0;
        int idx;
        for(int ii=0; ii < N; ++ii){ //which ii is the largest
            //mxl = max(mxl, dp[ii]);
            if(dp[ii]>mxl){
                mxl = dp[ii];
                idx=ii;
            }
        }

        vector<int> ret;
        while(idx!=-1){ //trace back the largest II for path
            ret.push_back(nums[idx]);
            idx=prev[idx];
        }

        return ret;
    }
};